## Numpy assignment

## 1 Preliminaries

In [1]:
# The usual preamble
import numpy as np

## Numpy

In [2]:
import numpy as np
x = np.array([2,3,1,0])
y =np.arange(10)
z = 2 + x

Write expressions in the next cell to retrieve 1 from `x`, 3 from `y`, and 3 from `z`.

In [3]:
print(x[2])
print(y[3])
print(z[2])

1
3
3


In the next cell, write an expression that generates a 5 by 5 array filled with zeros, and another  that generates a 1 by 1 array with a 1 in it.

In [4]:
np.zeros([5,5])

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [5]:
np.ones([1])

array([1.])

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  

In the next cell, write an expression that uses an assignment to a splice to make all the negative values in `a` be 0. 

In [7]:
a = np.arange(10,-1,-1) - 3
a

a[a < 0] = 0
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

One way to do this is:

In [8]:
a[-3:] = 0
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

But see if you can do it by writing an expression that make all negative values in `a`
be 0 no matter where they occur in `a`.  Hint: Use a Boolean conditipn

In [9]:
a = np.arange(10,-1,-1) - 3

Another option is to use `numpy` `where`, combined with an assignment.
The form is

```
>>> np.where(condition, x, y)
```

where `condition`, `x` and `y` are all arrays of the same size (or are broadcastable to the same size).
Return an array containing the members of `x` for which `condition` is `True`; otherwise return the corresponding member  of `y`.  It can be read "`x` where `condition`, else
`y`".

```
>>> a = np.array([1, 4.1, 9, 16])
>>> b = np.array([1,2,3,4])
>>> np.where((np.sqrt(a) - b) == 0,np.sqrt(a),b)
array([1., 2., 3., 4.])
```

In [10]:
a = np.arange(10,-1,-1) - 3
a = np.where(a < 0, 0, a)

In [11]:
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

In the next cell write an expression that produces an array containing result of multiplying 7 by each of the first 12 integers (1 - 12). There's a hard way to do this and an easy way.  The easy way uses elementwise operations.

In [12]:
7 * np.arange(1,13)

array([ 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84])

In [13]:
#elementwise operation 
np.multiply(7,np.arange(1,13))

array([ 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84])

## 2D arrays

In [14]:
import numpy as np
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In the next cell write an express that retrieves the last column of `room_matrix`.

In [15]:
room_matrix[:,3]

array([1, 2, 2, 1, 2])

In the next cell, write an expression that retrieves the third row of `room_matrix`.

In [16]:
room_matrix[:,2]

array([4, 3, 6, 3, 7])

In the next cell write an expression that retrieves the first two elements of the second column of `room_matrix`

In [17]:
room_matrix[:2,2]

array([4, 3])

In the next cell write an expression that retrieves the following submatrix from `room_matrix`:

```
array([[2,3],
       [3,6]])
```

In [18]:
room_matrix[1:3,[1,2]]

array([[2, 3],
       [3, 6]])

In the next cell, write an expression that returns a Boolean array that identifies the members of `room_matrix` which are greater than or equal to 7.

In [19]:
room_matrix >= 7

array([[False, False, False, False],
       [False, False, False, False],
       [ True, False, False, False],
       [False, False, False, False],
       [ True, False,  True, False]])

In the next cell, write an expression that produces an array containing the items in `room_matrix` which are greater than or equal to 7.  Note:  This question and the previous question do not have the same answer.

In [20]:
room_matrix[[room_matrix >= 7]]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  """Entry point for launching an IPython kernel.


array([ 8, 10,  7])

## Iris data exercise

In the next cell, use a Boolean mask to return all the rows of the iris data set whose column 3 value is greater than 1.8 (I mean the column whose **index** is 3, the last column).
The cell is not empty because you have been provided with the code which loads the data.
So you should keep that code and **add to it**.

In [21]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

In the next cell, use a Boolean array to return all the rows representing members of class 2 in the iris data set (the Boolean array is not the answer; it is used in retrieving the answer).

In [24]:
type(data)

sklearn.utils.Bunch

In [25]:

data.data[data.target == 2]

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.3, 2.9, 5.6, 1.8],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [4.9, 2.5, 4.5, 1.7],
       [7.3, 2.9, 6.3, 1.8],
       [6.7, 2.5, 5.8, 1.8],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [6.5, 3. , 5.5, 1.8],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6. , 2.2, 5. , 1.5],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.3, 2.7, 4.9, 1.8],
       [6.7, 3.3, 5.7, 2.1],
       [7.2, 3.2, 6. , 1.8],
       [6.2, 2.8, 4.8, 1.8],
       [6.1, 3. , 4.9, 1.8],
       [6.4, 2.8, 5.6, 2.1],
       [7.2, 3. , 5.8, 1.6],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [6.3, 2.8, 5.1, 1.5],
       [6.1, 2

Extra credit:  In the next cell write an expression -- or expressions -- which returns
all the members of class 2 in the data whose column 3 value is greater than 1.8.

Hint: Set a variable to the result of one of your answers to the last two questions.
That result is an array, and you can do more indexing on it.

In [27]:
Class_2 = data.data[data.target == 2]
Class_2[Class_2[:,2] > 1.18]

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.3, 2.9, 5.6, 1.8],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [4.9, 2.5, 4.5, 1.7],
       [7.3, 2.9, 6.3, 1.8],
       [6.7, 2.5, 5.8, 1.8],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [6.5, 3. , 5.5, 1.8],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6. , 2.2, 5. , 1.5],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.3, 2.7, 4.9, 1.8],
       [6.7, 3.3, 5.7, 2.1],
       [7.2, 3.2, 6. , 1.8],
       [6.2, 2.8, 4.8, 1.8],
       [6.1, 3. , 4.9, 1.8],
       [6.4, 2.8, 5.6, 2.1],
       [7.2, 3. , 5.8, 1.6],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [6.3, 2.8, 5.1, 1.5],
       [6.1, 2

Two Solutions.  The one not hinted at uses Boolean array conjunction.

Solution 2.  This one uses `&`.  Say you have X an array of numbers and you want a Boolean array
that tells you exactly which numbers are even (`X%2 == 0`) and greater than 4 (`X > 4`).  You can
combine these two conditions using `7`:

In [28]:
X = np.arange(15).reshape((3,5))
print(X)
(X%2 == 0) & (X > 4)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


array([[False, False, False, False, False],
       [False,  True, False,  True, False],
       [ True, False,  True, False,  True]])

To find solution 2, write a single expression that uses `&` which returns
all the rows of the Iris data class 2 in the data whose column 3 value is greater than 1.8.

This should produce a 2D array with 4 columns and 34 rows.

In [30]:
data.data[(data['target'] == 2) & (data.data[:,3] > 1.8)]

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.1],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [7.7, 3. , 6.1, 2.3],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.4, 2.1],
       [6.7, 3.1, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [5.8, 2.7, 5.1, 1.9],
       [6.8, 3.2, 5.9, 2.3],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5.2, 2.3],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3]])

## Boston Housing data exercise (UCI ML housing dataset)

In [31]:
from sklearn import datasets
boston = datasets.load_boston()
X,Y = boston.data,boston.target
print(X.shape, Y.shape)

(506, 13) (506,)


/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

The cell above loads the `sklearn` Boston housing data, which contains information about housing prices
for 501 houses sold in Boston, and the values of various attributes which have been found to be useful
in predicting home prices:

```
Index
0         - CRIM     per capita crime rate by town
1         - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
2         - INDUS    proportion of non-retail business acres per town
3         - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
4         - NOX      nitric oxides concentration (parts per 10 million)
5         - RM       average number of rooms per dwelling
6         - AGE      proportion of owner-occupied units built prior to 1940
7         - DIS      weighted distances to five Boston employment centres
8         - RAD      index of accessibility to radial highways
9         - TAX      full-value property-tax rate per $10,000
10        - PTRATIO  pupil-teacher ratio by town
11        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
12        - LSTAT    % lower status of the population
```

The array `X`, defined in the cell above, is a 501x13 array containing the values of all
these variables for 501 Boston houses.  

The information stored in Y, the value to be predicted is

```
       - MEDV     Median value of owner-occupied homes in $1000's
 
```

1.  Construct an expression or expressions that creates a new table that omits all houses whose tracts
    bound the Charles River.
2.  Construct an expression or expressions that creates a new table containing only the CRIM column and
    the ZN column.  This is  our new X.  Also create a new Y containing the pupil-teach ratio.
    This is to help someone research how to predict poor education conditions.  The shape of the new X should
    (506,2) and the shape of the new Y should be (506,).  Show that you checked this in your answer.

In [32]:
Bound_Data = boston.data[boston.data[:,3] !=1]
Bound_Data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [34]:
X = boston.data[:,0:2]
Y = boston.data[:,10]
print(X.shape)
print(Y.shape)

(506, 2)
(506,)


Write an exporession that computes how many housing tracts in the data set bound
the Charles River.

In [35]:
boston.data[boston.data[:,3] == 1].shape

(35, 13)

Checks!

## Part two

The answers for the following section are available on the inclass Boolean exercies solution notebook.

# Boolean combinations

In this section we load forest covertype data.  This is a multi class dataset that has data for 7 different forest covertypes (stored in the `target` attribute).  There are 581,012 forest plots with 54 attributes each (stored ina 581012x54 array) The first ten attributes are numerical, the last 44 are Boolean (true/false) attributes. Each of the Boolean attributes represents a qualitative soil type attribute which is either present or absent.  We will refer to all the attributes by their column index.  Thus the first attribute is attribute 0 and the last (a Boolean attribute) is attribute 53.

In [36]:
#from sklearn.datasets import load_wine
#wdata = load_wine()
from sklearn.datasets import fetch_covtype
data = fetch_covtype()
print(data.data.shape)  # data.data is the 581012 x 54 array
print(data.target.shape)     # data.target contains the class for each instance

(581012, 54)
(581012,)


Th first 26 columns of the plot index 50_210:

In [37]:
data.data[50_210,:26]

array([2.988e+03, 2.250e+02, 2.000e+00, 5.520e+02, 9.200e+01, 5.492e+03,
       2.160e+02, 2.410e+02, 1.620e+02, 4.290e+03, 1.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00])

The last 28 column of the plot index 50_210:

In [38]:
data.data[50_210,26:]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

[The Boolean arrays and masks notebook](https://github.com/gawron/python-for-social-science/blob/master/numpy/02_06_Boolean_Arrays_and_Masks.ipynb) discusses combining Boolean arrays with Boolean operators `&` (conceptually 'and') and `|` (conceptually 'or').   We can add to that `~` (conceptually 'not') to complete the list.  Look at the examples there.

 
The `sklearn` description of the covertype dataset is printed in the cell below. For a fuller understanding of the attributes in the dataset you might want to look at [the original UCI data set description.](https://archive.ics.uci.edu/ml/datasets/Covertype)

Forest covertypes
---------------------

The samples in this dataset correspond to 30×30m patches of forest in the US,
collected for the task of predicting each patch's cover type,
i.e. the dominant species of tree.
There are seven covertypes, making this a multiclass classification problem.
Each sample has 54 features, described on the
[dataset's homepage](https://archive.ics.uci.edu/ml/datasets/Covertype>).
Some of the features are boolean indicators
while others are discrete or continuous measurements.

**Data Set Characteristics:**

    =================   ============
    Classes                        7
    Samples total             581012
    Dimensionality                54
    Features                     int
    =================   ============

The function `sklearn.datasets.fetch_covtype` will load the covertype dataset;
it returns a dictionary-like object
with the feature matrix in the ``data`` member
and the target values in ``target``.
The dataset will be downloaded from the web if necessary.

`data.data` attribute is a `numpy` array containing the array.  As an example to help with this problem.  To compute an an array containing just the rows in which attribute 0 is greater than 50 you do. 

In [39]:
WW = data.data > 50
print(WW.shape)

(581012, 54)


The covertypes (or dominant tree) for each forest plot are in `data.target`, which is often the attribute used to store the classes in an `sklearn` clasification dataset.  There is data for 581,012 plots
so this is a 1D array of length 581,012.  

There are 7 distinct fiorest tyoes that a plot can have.

In [40]:
print(len(data.target))
print(set(data.target))

581012
{1, 2, 3, 4, 5, 6, 7}


### The problem

1. Write an expression that retrieves all rows that a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

2. Write an expression that retrieves the rows from either class 3 or class 5  that have a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

3. Write an expression that counts the rows from either class 3 or class 5  that have a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

NameError: ignored

In [44]:
#1
data.data[(data.data[:,1] > 300) & (data.data[:,12] ==1)]

array([[2780.,  346.,   13., ...,    0.,    0.,    0.],
       [2725.,  353.,   19., ...,    0.,    0.,    0.],
       [2747.,  329.,   19., ...,    0.,    0.,    0.],
       ...,
       [2619.,  336.,   13., ...,    0.,    0.,    0.],
       [2630.,  317.,   12., ...,    0.,    0.,    0.],
       [2633.,  309.,    9., ...,    0.,    0.,    0.]])

In [45]:
#2
data.data[(data.target ==3|5) & (data.data[:,1] > 300) & (data.data[:,12] == 1)]

array([[3.399e+03, 3.030e+02, 5.000e+00, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [3.410e+03, 3.100e+02, 7.000e+00, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [3.413e+03, 3.030e+02, 8.000e+00, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       ...,
       [3.305e+03, 3.510e+02, 1.400e+01, ..., 0.000e+00, 0.000e+00,
        0.000e+00],
       [3.308e+03, 3.010e+02, 1.000e+01, ..., 0.000e+00, 0.000e+00,
        0.000e+00],
       [3.313e+03, 3.310e+02, 6.000e+00, ..., 0.000e+00, 0.000e+00,
        0.000e+00]])

In [46]:
#3
#Column and Row count
data.data[(data.target == 3|5) & (data.data[:,1] > 300) & (data.data[:,12] ==1)].shape

(2774, 54)

In [47]:
# just row count 
len(data.data[(data.target == 3|5) & (data.data[:,1] > 300) & (data.data[:,12] ==1)])

2774